[jit] Give functions qualified names

suo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #22504 [WIP] classtypes hold weak_ptr
* #22692 [jit] Make traced fns also go into the global python CU
* #22221 [jit] _script_compile and _script_class_compile add to the python CU
* #22667 [jit] make CompilationUnit::define return defined functions
* #22207 [jit] refactor self to be a class again
* **#22206 [jit] Give functions qualified names**

This makes it so that all code objects owned by a CompilationUnit will
have a qualified name.

This also means that Module performs one new duty: it does
qualified name resolution by concating the Module qualified name with
the method name to find the method in the owning CU.

Differential Revision: [D15998762](https://our.internmc.facebook.com/intern/diff/D15998762)

suo committed:
[jit] Give functions qualified names

This makes it so that all code objects owned by a CompilationUnit will
have a qualified name.

This also means that Module performs one new duty: it does
qualified name resolution by concating the Module qualified name with
the method name to find the method in the owning CU.

suo committed:
Update on "[jit] Give functions qualified names"

[jit] Give functions qualified names

This makes it so that all code objects owned by a CompilationUnit will
have a qualified name.

This also means that Module performs one new duty: it does
qualified name resolution by concating the Module qualified name with
the method name to find the method in the owning CU.

gh-metadata: pytorch pytorch 22206 gh/suo/70/head

suo committed:
Update on "[jit] Give functions qualified names"

[jit] Give functions qualified names

This makes it so that all code objects owned by a CompilationUnit will
have a qualified name.

This also means that Module performs one new duty: it does
qualified name resolution by concating the Module qualified name with
the method name to find the method in the owning CU.

gh-metadata: pytorch pytorch 22206 gh/suo/70/head

suo committed:
Update on "[jit] Give functions qualified names"

[jit] Give functions qualified names

This makes it so that all code objects owned by a CompilationUnit will
have a qualified name.

This also means that Module performs one new duty: it does
qualified name resolution by concating the Module qualified name with
the method name to find the method in the owning CU.

gh-metadata: pytorch pytorch 22206 gh/suo/70/head

zdevito commented on torch/csrc/jit/script/module.h:
```diff
@@ -75,7 +75,7 @@ struct TORCH_API Method {
   }
 
-  const std::string& name() const {
+  std::string name() const {
```
Why did this change? Is the result different? I assume this would be the unqualified name (since methods really don't have qualified names).

suo replied:
Yeah, I think this was left over from something. Will change

zdevito commented on torch/csrc/jit/script/class_type.cpp:
```diff
@@ -8,7 +8,8 @@ namespace c10 {
 
 Function* ClassType::getMethod(const std::string& name) const {
-  return compilation_unit_->find_function(name);
+  const auto qualname = QualifiedName(*qualified_name_obj(), name);
```
qualified_name_obj is really confusing. If you look at the methods in NamedType they are pretty redundant. We really should have NamedType just return the qualified name and rename qualified_name_obj to just qualifiedName.

zdevito commented on torch/csrc/jit/script/compilation_unit.h:
```diff
@@ -173,29 +181,31 @@ struct TORCH_API CompilationUnit {
 
  private:
   std::unique_ptr<Function> define(
+      const c10::QualifiedName& name,
```
I think this should be prefix, otherwise it is denormalized with Def

zdevito commented on torch/csrc/jit/script/module.h:
```diff
@@ -222,8 +226,9 @@ struct TORCH_API Module {
   }
-  c10::optional<Method> find_method(const std::string& name) const {
-    if (const auto fn = class_compilation_unit()->find_function(name)) {
+  c10::optional<Method> find_method(const std::string& basename) const {
```
This runs in the user's eval loop, so I am not really happy about having to allocate a new string in order to look up the function. I feel like the ClassType should just be holding onto this mapping directly.

zdevito reviewed (approved):
This looks pretty good. The one thing that really should be changed is that `define` really should not take two definitions for the name (qualName and name in def), and ignore the one in the def. I think we should just add the prefix as part of the define function.

suo committed:
Update on "[jit] Give functions qualified names"

[jit] Give functions qualified names

This makes it so that all code objects owned by a CompilationUnit will
have a qualified name.

This also means that Module performs one new duty: it does
qualified name resolution by concating the Module qualified name with
the method name to find the method in the owning CU.

gh-metadata: pytorch pytorch 22206 gh/suo/70/head

suo committed:
Update on "[jit] Give functions qualified names"

[jit] Give functions qualified names

This makes it so that all code objects owned by a CompilationUnit will
have a qualified name.

This also means that Module performs one new duty: it does
qualified name resolution by concating the Module qualified name with
the method name to find the method in the owning CU.

gh-metadata: pytorch pytorch 22206 gh/suo/70/head

suo committed:
Update on "[jit] Give functions qualified names"

[jit] Give functions qualified names

This makes it so that all code objects owned by a CompilationUnit will
have a qualified name.

This also means that Module performs one new duty: it does
qualified name resolution by concating the Module qualified name with
the method name to find the method in the owning CU.

gh-metadata: pytorch pytorch 22206 gh/suo/70/head

suo committed:
Update on "[jit] Give functions qualified names"

[jit] Give functions qualified names

This makes it so that all code objects owned by a CompilationUnit will
have a qualified name.

This also means that Module performs one new duty: it does
qualified name resolution by concating the Module qualified name with
the method name to find the method in the owning CU.

gh-metadata: pytorch pytorch 22206 gh/suo/70/head

suo commented on torch/csrc/jit/script/class_type.cpp:
```diff
@@ -63,7 +64,17 @@ size_t ClassType::addAttribute(
 
 std::vector<Function*> ClassType::methods() const {
-  return compilation_unit()->get_functions();
+  // TODO: this needs to be made more efficient!
```
I will move that stuff from the other PR to this one.

suo committed:
Update on "[jit] Give functions qualified names"

[jit] Give functions qualified names

This makes it so that all code objects owned by a CompilationUnit will
have a qualified name.

This also means that Module performs one new duty: it does
qualified name resolution by concating the Module qualified name with
the method name to find the method in the owning CU.

gh-metadata: pytorch pytorch 22206 gh/suo/70/head

suo committed:
Update on "[jit] Give functions qualified names"

[jit] Give functions qualified names

This makes it so that all code objects owned by a CompilationUnit will
have a qualified name.

This also means that Module performs one new duty: it does
qualified name resolution by concating the Module qualified name with
the method name to find the method in the owning CU.

gh-metadata: pytorch pytorch 22206 gh/suo/70/head
